### PR TITLE
Generator of natural numbers yields IllegalStateException

### PR DESCRIPTION
**Pull request by JaroslavTulach**
[ci no changelog needed]

This PR provides a test and a fix for following problem:

### Steps to reproduce

Trying to write _sieve of erathostenes_ in the Enso language. Step no. 1 is to get a decent generator of numbers from `2` to infinity. 

```
from Standard.Base import IO

type Generator h t

natural =
    gen ~n = Generator n (gen n+1)
    gen 2

main = case here.natural of
    Generator v _ -> IO.println v
```

Example https://github.com/jtulach/sieve/commit/9c1f2cf201a6246530aaab439a0a359a4cde3de8

### Expected

The above program might be the right way to generate the endless list. Ideally it shall print `2`.

Regardless whether the program is correct or not, it  certainly it shall not fail with engine error as it does with engine build on tag `2022.1.1-nightly.2022-05-07`.

### Actual

```
built-distribution/enso-engine-0.0.0-dev-linux-amd64/enso-0.0.0-dev/bin/enso --run sieve.enso
Execution finished with an error: java.lang.IllegalStateException
        at <java> org.graalvm.truffle/com.oracle.truffle.api.frame.FrameUtil.getObjectSafe(FrameUtil.java:61)
        at <java> org.enso.interpreter.node.scope.ReadLocalVariableNode.readGeneric(ReadLocalVariableNode.java:63)
        at <java> org.enso.interpreter.node.scope.ReadLocalVariableNodeGen.executeGeneric(ReadLocalVariableNodeGen.java:50)
        at <java> org.enso.interpreter.node.callable.ApplicationNode.evaluateArguments(ApplicationNode.java:83)
        at <java> org.enso.interpreter.node.callable.ApplicationNode.executeGeneric(ApplicationNode.java:100)
        at <java> org.enso.interpreter.node.ClosureRootNode.execute(ClosureRootNode.java:72)
        at <enso> argument<0>(Internal)
        at <enso> argument<0>(sieve.enso:6:31-33)
        at <enso> argument<0>(sieve.enso:6:31-33)
        at <enso> argument<0>(sieve.enso:6:31-33)
        at <enso> gen(sieve.enso:6:14-34)
        at <enso> argument<1>(sieve.enso:6:27-33)
        at <enso> gen(sieve.enso:6:14-34)
        at <enso> argument<1>(sieve.enso:6:27-33)
        at <enso> gen(sieve.enso:6:14-34)
        at <enso> argument<1>(sieve.enso:6:27-33)
        at <enso> gen(sieve.enso:6:14-34)
        at <enso> argument<1>(sieve.enso:6:27-33)
        at <enso> gen(sieve.enso:6:14-34)
        at <enso> sieve.natural(sieve.enso:7:5-9)
        at <enso> sieve.main(sieve.enso:9:13-24)
```

### CODE REVIEW

**Review by hubertp — APPROVED** [review]
I think there is a common pattern emerging for such tests cases. Might want to refactor it for DRY in the future. But LGTM at this point.
Not sure if the CHANGELOG doesn't need an entry, since this is not a regression, though.

File: engine/runtime/src/main/java/org/enso/interpreter/node/scope/ReadLocalVariableNode.java [hunk 1]
```diff
@@ -60,7 +60,7 @@ protected Object readGeneric(VirtualFrame frame) {
     if (getFramePointer().getParentLevel() == 0)
       return FrameUtil.getObjectSafe(frame, getFramePointer().getFrameSlot());
     MaterializedFrame currentFrame = getProperFrame(frame);
-    return FrameUtil.getObjectSafe(currentFrame, getFramePointer().getFrameSlot());
+    return currentFrame.getValue(getFramePointer().getFrameSlot());
```
Comment (kustosz) [inline engine/runtime/src/main/java/org/enso/interpreter/node/scope/ReadLocalVariableNode.java:71]: Please confirm this does not cause a slowdown. To my knowledge `getValue` is _way_ slower than `getObjectSafe`.

File: engine/runtime/src/test/java/org/enso/interpreter/test/LazyGeneratorTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    type Generator h t
+
+    natural =
+        gen ~n = if (n >= 100) then this else Generator n (@Tail_Call gen n+1)
+        gen 2
+                                                          
+    Generator.n = case this of
+        Generator n _ -> n
+    Generator.next = case this of
+        Generator _ n -> n
```
Comment (radeusgd) [inline engine/runtime/src/test/java/org/enso/interpreter/test/LazyGeneratorTest.java]: Why not just put this test in `tests/Test`? Does it in any way rely on the special setup and cannot be run as a regular Enso script?

**Comment by JaroslavTulach** [comment]
The current code seems to cause regression in `NamedDefaultedArgumentBenchmarks`:
```xml
         <case>
             <label>org.enso.interpreter.bench.benchmarks.semantic.NamedDefaultedArgumentBenchmarks.benchSumTCOWithDefaultArgs</label>
             <scores>
                 <score>50.31606527288427</score>
+                <score>455.87280858695647</score>
             </scores>
         </case>
         <case>
             <label>org.enso.interpreter.bench.benchmarks.semantic.NamedDefaultedArgumentBenchmarks.benchSumTCOWithNamedArgs</label>
             <scores>
                 <score>50.27513270669849</score>
+                <score>461.69807143122534</score>
             </scores>
```

File: engine/runtime/src/main/java/org/enso/interpreter/node/scope/ReadLocalVariableNode.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/node/scope/ReadLocalVariableNode.java:71]: Leaving this specialization unchanged, but rewriting to `getValue` on type slot exception as of 3d9bc2714e2f7

File: engine/runtime/src/test/java/org/enso/interpreter/test/LazyGeneratorTest.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/interpreter/test/LazyGeneratorTest.java]: Providing as a regular test as of 5dbbe9c427020bce6771e44c012c1345baa012a4

**Comment by JaroslavTulach** [comment]
[3d9bc27](https://github.com/enso-org/enso/pull/3440/commits/3d9bc2714e2f7ce997baac3ba37921e743378ccb) fixes many benchmarks, but `AtomBenchmarks.benchMapReverseCurryList` and `AtomBenchmarks.benchMapReverseList` and `AtomBenchmarks.benchReverseListMethods` are oscillating. However these benchmarks seem to be [unstable even in `develop` branch](https://www.pivotaltracker.com/n/projects/2539304/stories/182199750/comments/231288210) and this PR doesn't seem to affect their _unstability_ (e.g. sometimes each of them can run fast).

What do you think? I need one more approval.

**Review by 4e6 — APPROVED** [review]

**Pull request merged**